advanced_source/torch_script_parallelism.rst

```diff
@@ -0,0 +1,276 @@
+Dynamic Parallelism in TorchScript
+==================================
+
+In this tutorial, we introduce the syntax for doing *dynamic inter-op parallelism*
+in TorchScript. This parallelism has the following properties:
+
+* dynamic - The number of parallel tasks created and their workload can depend on the control flow of the program.
+* inter-op - The parallelism is concerned with running TorchScript program fragments in parallel. This is distinct from *intra-op parallelism*, which is concerned with splitting up individual operators and running subsets of the operator's work in parallel.
+Basic Syntax
+------------
+
+The two important APIs for dynamic parallelism are:
+
+* ``torch.jit.fork(fn : Callable, *args, **kwargs) -> torch.jit.Future[T]``
+* ``torch.jit.wait(fut : torch.jit.Future[T]) -> T``
+
+A good way to demonstrate how these work is by way of an example:
+
+.. code-block:: python
+
+    import torch
+
+    def foo(x):
+        return torch.neg(x)
+
+    @torch.jit.script
+    def example(x):
+        # Call `foo` normally
+        x_normal = foo(x)
+
+        # Call `foo` using parallelism:
+        # First, we "fork" off a task. This task will run `foo` with argument `x`
+        future = torch.jit.fork(foo, x)
+        # Second, we "wait" on the task. Since the task may be running in
+        # parallel, we have to "wait" for its result to become available.
+        x_parallel = torch.jit.wait(future)
+
+        return x_normal, x_parallel
+
+    print(example(torch.ones(1))) # (-1., -1.)
+
+
+``fork()`` takes the callable ``fn`` and arguments to that callable ``args``
+and ``kwargs`` and creates an asynchronous task for the execution of ``fn``.
+``fn`` can be a function, method, or Module instance. ``fork()`` returns a
+reference to the value of the result of this execution, called a ``Future``.
+Because ``fork`` returns immediately after creating the async task, ``fn`` may
+not have been executed by the time the line of code after the ``fork()`` call
+is executed. Thus, ``wait()`` is used to wait for the async task to complete
+and return the value.
+
+These constructs can be used to overlap the execution of statements within a
+function (shown in the worked example section) or be composed with other language
+constructs like loops:
+
+.. code-block:: python
+
+    import torch
+    from typing import List
+
+    def foo(x):
+        return torch.neg(x)
+
+    @torch.jit.script
+    def example(x):
+        futures = torch.jit.annotate(List[torch.jit.Future[torch.Tensor]], [])
+        for _ in range(100):
+            futures.append(torch.jit.fork(foo, x))
+
+        results = []
+        for future in futures:
+            results.append(torch.jit.wait(future))
+
+        return torch.sum(torch.stack(results))
+
+    print(example(torch.ones([])))
+
+.. note::
+
+    When we initialized an empty list of Futures, we needed to add an explicit
+    type annotation to the value via ``torch.jit.annotate``. This is because
```

Comment:
similarly, maybe remove torch.jit.annotate here. We should probably have an example using list comprehensions, bc  it's easier without the type annotation. `futures = [torch.jit.fork(foo, x) for _ in range(100)]`.  

I think this whole section could be a lot shorter, something like. `Empty containers default to assuming Tensor values, so we annotate the list constructor as being of type `List[torch.jit.Future[Tensor]'`. And could maybe in inline in the example.